python/tank/bootstrap/threading.py

```diff
@@ -0,0 +1,269 @@
+        self._engine_name = engine_name
+        self._entity = entity
+
+        self._progress_callback = None
+        self._completed_callback = None
+        self._failed_callback = None
+
+        # Create a worker that can bootstrap asynchronously in a background thread.
+        self._worker = BootstrapWorker(toolkit_manager, engine_name, entity)
+
+        # This QThread object will live in the main thread, not in the new thread it will manage.
+        self._thread = QtCore.QThread()
+
+        # Make the worker operate with the new thread affinity and use the QThread object event loop.
+        self._worker.moveToThread(self._thread)
+
+        # Start to work when the QThread object will have started its event loop in its new thread context.
+        self._thread.started.connect(self._worker.work)
+
+        # Make the worker report on the toolkit bootstrap progress.
+        self._worker.progressing.connect(self._progress_bootstrap)
+
+        # Handle completion of the toolkit bootstrap by the worker.
+        self._worker.completed.connect(self._complete_bootstrap)
+
+        # Handle failure of the toolkit bootstrap by the worker.
+        self._worker.failed.connect(self._fail_bootstrap)
+
+    def set_callbacks(self, progress_callback, completed_callback, failed_callback):
+        """
+        Sets the callback functions used to signify the progress, the completion and the failure of the bootstrap.
+
+        These functions must have the following signatures:
+
+            ``progress_callback(message, current_index, maximum_index)``
+
+            where:
+            - ``message`` is the progress message to report.
+            - ``current_index`` is an optional current item number being looped over.
+            - ``maximum_index`` is an optional maximum item number being looped over.
+`
+            ``completed_callback(engine)``
+
+            where:
+            - ``engine``is the launched :class:`sgtk.platform.Engine` instance.
+
+            ``failed_callback(step, exception)``
+
+            where:
+            - ``step``is the bootstrap step ("sgtk" or "engine") that raised the exception.
+            - ``exception``is the python exception raised while bootstrapping.
+
+        :param progress_callback: Callback function that reports back on the toolkit and engine bootstrap progress.
+        :param completed_callback: Callback function that handles cleanup after successful completion of the bootstrap.
+        :param failed_callback: Callback function that handles cleanup after failed completion of the bootstrap.
+        """
+
+        self._progress_callback = progress_callback
+        self._completed_callback = completed_callback
+        self._failed_callback = failed_callback
+
+    def bootstrap(self):
+        """
+        Starts the asynchronous bootstrap logic.
+        """
+
+        # Start the QThread object event loop in its new thread context.
+        self._thread.start()
+
+    @QtCore.Slot()
```

Comment:
Are the decorators necessary? I don' think we use these anywhere else in tk code.